src/Storages/MergeTree/ReplicatedMergeTreeRestartingThread.cpp

```diff
@@ -339,10 +339,15 @@ void ReplicatedMergeTreeRestartingThread::shutdown(bool part_of_full_shutdown)
     /// Stop restarting_thread before stopping other tasks - so that it won't restart them again.
     need_stop = true;
     task->deactivate();
+
+    /// Explicitly set the event, because the restarting thread will not set it again
+    if (part_of_full_shutdown)
+        storage.startup_event.set();
+
     LOG_TRACE(log, "Restarting thread finished");
 
-    /// Stop other tasks.
-    partialShutdown(part_of_full_shutdown);
+    setReadonly(part_of_full_shutdown);
+
```

Comment:
Looks like `setReadonly` will be called twice: from `ReplicatedMergeTreeRestartingThread::shutdown` from `flushAndPrepareForShutdown` and then again from `ReplicatedMergeTreeRestartingThread::partialShutdown` from `StorageReplicatedMergeTree::shutdown`. On the second call, `became_readonly` will be false here:
https://github.com/ClickHouse/ClickHouse/blob/2924837c7e6030058b87b9a65a06276543430cfd/src/Storages/MergeTree/ReplicatedMergeTreeRestartingThread.cpp#L353-L355

So we will decrement the metric:
https://github.com/ClickHouse/ClickHouse/blob/2924837c7e6030058b87b9a65a06276543430cfd/src/Storages/MergeTree/ReplicatedMergeTreeRestartingThread.cpp#L360-L365